keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -323,11 +338,185 @@ myOtherField: newly added field
+```go
+// requireStructuralSchema returns true if schemas specified must be structural
+func requireStructuralSchema(oldCRDSpec *apiextensions.CustomResourceDefinitionSpec) bool {
+	if oldCRDSpec != nil && specHasNonStructuralSchema(oldCRDSpec) {
+		// don't tighten validation on existing persisted data
+		return false
+	}
+	return true
+}
+```
+
+It is not possible to generically allow this weaker form of ratcheting on all
+types of validations - it may allow data that would have failed a validation
+on an older schema.
+
+To enable schemas to add validations with more complicated ratcheting logic
+than a DeepEqual check (e.g. to allow a new `minLength` restriction to allow any
+value as long as it is getting closer the new minimum); we turn to CEL.
+
+#### Ratcheting Rules in CEL
+
+CRD Validation Rules currently have two types:
+
+- Normal Rule (Applies to `CREATE` and `UPDATE` requests)
+- Transition Rule (Applies to `UPDATE` requests only)
+
+With only these rule types we are not able to express any logic of the form 
+"old values must obey one schema, and new values another".
+
+This is way we propose to add a new rule type to enable this functionality for
+ratcheting:
+
+- Create-Only Rule (Applies to `CREATE` requests)
+
+This would involve the following change to the `ValidationRule` type:
+
+```go
+// ValidationRule describes a validation rule written in the CEL expression language.
+type ValidationRule struct {
+  // ...
+	Rule string `json:"rule" protobuf:"bytes,1,opt,name=rule"`
+	Message string `json:"message,omitempty" protobuf:"bytes,2,opt,name=message"`
+	MessageExpression string `json:"messageExpression,omitempty" protobuf:"bytes,3,opt,name=messageExpression"`
+	Reason *FieldValueErrorReason `json:"reason,omitempty" protobuf:"bytes,4,opt,name=reason"`
+	FieldPath string `json:"fieldPath,omitempty" protobuf:"bytes,5,opt,name=fieldPath"`
+  // ...
+
+  // New
+  // Request types that this validation rule applies to
+  // If empty or unset, will apply to all applicable operations for the
+  // given rule expression.
+  Operations []OperationType `json:"modality,omitempty"`
+}
+
+// OperationType specifies an operation for a request.
+// +enum
+type OperationType string
+
+const (
+	Create       OperationType = "CREATE"
+	Update       OperationType = "UPDATE"
+)
+```
+
+With this feature in place, users are able to express complicating ratcheting
+logic in CEL, to express racheting conditions based on values they know
+their controllers support:
+
+
+```yaml
+x-kubernetes-validations:
+- rule: len(self) >= 5
+  operations: [Create]
+  message: New objects need names longer or equal to 5.
+- rule: len(self) >= 5 || len(oldSelf) < len(self)
+  operations: [Create, Update]
+  message: Names shorter than 5 are not allowed anymore. Any changes to the name must make the name longer
```

Comment:
That rule could technically be an update only then? (it's sort of inferred from being a transition rule but kind of confusing because it says create).

Reply:
Alternatively, I think you could have made a type of rule that is marked as `deprecated`, allows some types of updates to the object, automatically sends a warning, automatically counts occurences of updates to deprecated fields, can count number of objects in etcd that have deprecated fields, can be time-based disabled etc (deprecated value hasn't been used in X time, disable it automatically). I think basically what I'm saying is that if the right deprecated semantic was in there, we would have a lot more power to build additional action on it.

Reply:
ah this was a typo in the KEP. Changed to `[Update]`. It would be a validation error to use `oldSelf` in a CREATE rule